api/src/services/replication-limit-log.js

```diff
@@ -0,0 +1,93 @@
+const moment = require('moment');
+const db = require('../db');
+const logger = require('../logger');
+
+const LOG_TYPES = {
+  replicationCount: 'replication-count-'
+};
+const LOG_COUNT_DIFF = 100;
+const LOG_MONTH_DIFF = 1;
+
+const persistLog = (info) => {
+  if (!info || !info.user || !info.user.user_name) {
+    const error = new Error('Error when persisting log: Log Information missing.');
+    return Promise.reject(error);
+  }
+
+  const logDocId = LOG_TYPES.replicationCount + info.user.user_name;
+
+  return db.medicLogs
+    .get(logDocId)
+    .catch(error => {
+      if (error.status === 404) {
+        return { _id: logDocId };
+      }
+      throw error;
+    })
+    .then(doc => {
+      if (!isLogDifferent(doc, info)) {
+        return;
+      }
+
+      const logDoc = Object.assign(doc, info);
+      return db.medicLogs.put(logDoc);
+    });
+};
+
+const isLogDifferent = (oldLog, newLog) => {
+  if (!oldLog.log_date && !oldLog.count) {
+    return true;
+  }
+
+  const oldLogDate = moment(oldLog.log_date);
+  const newLogDate = moment(newLog.log_date);
+  const monthDiff = newLogDate.diff(oldLogDate, 'months', true);
+  const countDiff = Math.abs(oldLog.count - newLog.count);
+
+  return monthDiff > LOG_MONTH_DIFF || countDiff > LOG_COUNT_DIFF;
+};
+
+const getLogsByType = (docPrefix) => {
+  const options = {
+    startkey: docPrefix,
+    endkey: docPrefix + '\ufff0',
+    include_docs: true
+  };
+
+  return db.medicLogs
+    .allDocs(options)
+    .then((result) => result.rows.map(row => row.doc));
```

Comment:
I'm a little worried about performance here. If we have 10k users this will fetch 10k docs every time the monitoring api is called.

I think we should add another exported function for just getting the count then we can optimise this path.

One way to optimise it is to use a map reduce view which emits the doc if it's over the limit. Then you can use the reduce aggregation functions to get CouchDB to `_count` the number of docs emitted so the heavily lifting is all done in CouchDB and API doesn't need to do anything.

Does that make sense?

Reply:
I'm going to research this one in CouchDB, haha, I'll get back on this one.

Reply:
Done!

Reply:
Done!